openforcefield/typing/engines/smirnoff/parameters.py

```diff
@@ -2219,12 +2224,78 @@ def create_force(self, system, topology, **kwargs):
                 len(improper_matches)))
 
 
-class vdWHandler(ParameterHandler):
+class _NonbondedHandler(ParameterHandler):
+    """Base class for ParameterHandlers that deal with OpenMM NonbondedForce objects."""
+    _OPENMMTYPE = openmm.NonbondedForce
+
+    def create_force(self, system, topology, **kwargs):
+        # If we aren't yet keeping track of which molecules' charges have been assigned by which charge methods,
+        # initialize a dict for that here.
+        # TODO: This should be an attribute of the _system_, not the _topology_. However, since we're still using
+        #  OpenMM's System class, I am storing this data on the OFF Topology until we make an OFF System class.
+        if not hasattr(topology, "_ref_mol_to_charge_method"):
+            topology._ref_mol_to_charge_method = {ref_mol: None for ref_mol in topology.reference_molecules}
+
+        # Retrieve the system's OpenMM NonbondedForce
+        existing = [system.getForce(i) for i in range(system.getNumForces())]
+        existing = [f for f in existing if type(f) == self._OPENMMTYPE]
+
+        # If there isn't yet one, initialize it and populate it with particles
+        if len(existing) == 0:
+            force = self._OPENMMTYPE()
+            system.addForce(force)
+            # Create all particles.
+            for _ in topology.topology_particles:
+                force.addParticle(0.0, 1.0, 0.0)
+        else:
+            force = existing[0]
+
+        return force
+
+    def mark_charges_assigned(self, ref_mol, topology):
+        """
+        Record that charges have been assigned for a reference molecule.
+
+        Parameters
+        ----------
+        ref_mol : openforcefield.topology.Molecule
+            The molecule to mark as having charges assigned
+        topology : openforcefield.topology.Topology
+            The topology to record this information on.
+
+        """
+        # TODO: Change this to interface with system object instead of topology once we move away from OMM's System
+        topology._ref_mol_to_charge_method[ref_mol] = self.__class__
```

Comment:
The `_ref_mol_to_charge_method` dict records where each unique molecule in the topology gets its charge from. It's initialized to have them all be `None`, but those get filled in as charging occurs.

Since `_NonbondedForce` is inherited by `ElectrostaticsHandler`, `ToolkitAM1BCCHandler`, and `LibraryChargeHandler`, when `self.__class__` actually runs, it will be inside of one of those classes. So the dict will be populated with pairs of the form {`reference_molecule`: charge-assigning `_NonbondedHandler`-derived class}.